File: backend/common/roles.py
from django.db import models


class Role(models.TextChoices):
    OWNER = "owner", "Owner"
    ADMIN = "admin", "Admin"
    AGENT = "agent", "Agent"
    VIEWER = "viewer", "Viewer"


ROLE_RANK = {Role.OWNER: 40, Role.ADMIN: 30, Role.AGENT: 20, Role.VIEWER: 10}


def role_at_least(role: str, minimum: str) -> bool:
    """True when ``role`` is ``minimum`` or higher (owner > admin > agent > viewer)."""
    return ROLE_RANK.get(role, 0) >= ROLE_RANK[minimum]

File: frontend/src/lib/roles.ts
export type Role = 'owner' | 'admin' | 'agent' | 'viewer'

const rank: Record<Role, number> = { viewer: 0, agent: 1, admin: 2, owner: 3 }

export const roleLabels: Record<Role, string> = {
  owner: 'Owner',
  admin: 'Admin',
  agent: 'Agent',
  viewer: 'Viewer',
}

/** True when `role` is at least `minRole` (owner > admin > agent > viewer). */
export function hasRole(role: Role | null | undefined, minRole: Role = 'viewer'): boolean {
  if (!role) return false
  return rank[role] >= rank[minRole]
}
